go/client/cmd_inbox.go

```diff
@@ -0,0 +1,156 @@
+// Copyright 2015 Keybase, Inc. All rights reserved. Use of
+// this source code is governed by the included BSD license.
+
+package client
+
+import (
+	"encoding/hex"
+	"strings"
+
+	"golang.org/x/net/context"
+
+	"github.com/keybase/cli"
+	"github.com/keybase/client/go/libcmdline"
+	"github.com/keybase/client/go/libkb"
+	keybase1 "github.com/keybase/client/go/protocol"
+	"github.com/keybase/gregor/protocol/chat1"
+	"github.com/keybase/gregor/protocol/gregor1"
+)
+
+type uidUsernameMapper map[keybase1.UID]libkb.NormalizedUsername
+
+func (m uidUsernameMapper) getUsername(g *libkb.GlobalContext, uid keybase1.UID) (username libkb.NormalizedUsername, err error) {
+	if m == nil {
+		m = make(uidUsernameMapper)
+	}
+
+	if username, ok := m[uid]; ok {
+		return username, nil
+	}
+
+	loadUserArg := libkb.NewLoadUserArg(g)
+	loadUserArg.UID = uid
+
+	var ret *libkb.User
+	if ret, err = libkb.LoadUser(loadUserArg); err != nil {
+		return username, err
+	}
+
+	m[uid] = ret.GetNormalizedName()
+	return username, err
+}
+
+type cmdInbox struct {
+	libkb.Contextified
+	chatLocalClient keybase1.ChatLocalInterface // for testing only
+}
+
+func newCmdInbox(cl *libcmdline.CommandLine, g *libkb.GlobalContext) cli.Command {
+	return cli.Command{
+		Name:         "inbox",
+		Usage:        "Show new messages in inbox",
+		ArgumentHelp: "",
+		Action: func(c *cli.Context) {
+			cl.ChooseCommand(&cmdInbox{Contextified: libkb.NewContextified(g)}, "inbox", c)
+		},
+	}
+}
+
+func (c *cmdInbox) getMessagesFlattened() (messages cliChatMessages, err error) {
+	chatClient := c.chatLocalClient // should be nil unless in test
+	if chatClient == nil {
+		chatClient, err = GetChatLocalClient(c.G())
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	var mapper uidUsernameMapper
+
+	ipagination := &chat1.Pagination{}
+getinbox:
+	for {
+		ipagination.Num = 32
+		iview, err := chatClient.GetInboxLocal(context.TODO(), ipagination)
+		if err != nil {
+			return messages, err
+		}
+
+		tpagination := &chat1.Pagination{}
+	getthread:
+		for _, conv := range iview.Conversations {
+			tpagination.Num = 32
+			tview, err := chatClient.GetThreadLocal(context.TODO(), keybase1.GetThreadLocalArg{
+				ConversationID: conv.Metadata.ConversationID,
+				Pagination:     tpagination,
+			})
+			if err != nil {
+				return nil, err
+			}
+
+			for _, m := range tview.Messages {
+				if len(m.MessagePlaintext.MessageBodies) == 0 {
+					continue
+				}
+
+				var body string
+				if m.MessagePlaintext.MessageBodies[0].Type == chat1.MessageType_TEXT {
+					body = formatChatText(m.MessagePlaintext.MessageBodies[0].Text)
+				} else if m.MessagePlaintext.MessageBodies[0].Type == chat1.MessageType_ATTACHMENT {
+					body = formatChatAttachment(m.MessagePlaintext.MessageBodies[0].Attachment)
+				} else {
+					continue
```

Comment:
What do you think we should do for `chat1.MessageType_GENERIC_DATA` to allow applications to talk to each other? A separate CLI subcommand or modifications to this one?